app_build_suite/build_steps/build_step.py

```diff
@@ -0,0 +1,212 @@
+        """
+        return self.__class__.__name__
+
+    @property
+    @abstractmethod
+    def steps_provided(self) -> List[StepType]:
+        """
+        This defines types of steps this BuildStep should be executed for. If a user filters the set of steps
+        and the steps listed here don't match any of the steps selected by the user, the whole BuildStep
+        won't be executed for this run.
+        :return: Returns a list with elements from ALL_STEPS.
+        """
+        raise NotImplementedError
+
+    def initialize_config(self, config_parser: configargparse.ArgParser) -> None:
+        """
+        Provide configuration options supported by this BuildStep. Needs to only act on ArgParser and can't
+        run any blocking/long operations.
+        :param config_parser: configargparse.ArgParser to add the configuration options to.
+        :return: None
+        """
+        pass
+
+    def pre_run(self, config: argparse.Namespace) -> None:
+        """
+        Execute any pre-run validation or assertion logic.
+        :param config: Ready (parsed) configuration Namespace object.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def run(self, config: argparse.Namespace, context: Dict[str, Any]) -> None:
+        """
+        Execute actual build action of the BuildStep.
+        :param context: A context where different components can save data to share with other components.
+        :param config: Ready (parsed) configuration Namespace object.
+        :return: None
+        """
+        raise NotImplementedError
+
+    def cleanup(
+        self,
+        config: argparse.Namespace,
+        context: Dict[str, Any],
+        has_build_failed: bool,
+    ) -> None:
+        """
+        Clean up any resources used during the BuildStep.
+        :param context: A context where different components can save data to share with other components.
+        :param has_build_failed: A boolean set to True if the cleanup is run after any of the BuildSteps
+        failed their run step
+        :param config: Ready (parsed) configuration Namespace object.
+        :return: None
+        """
+        pass
+
+    def _assert_binary_present_in_path(self, bin_name: str) -> None:
+        """
+        Checks if binary is available in the system. Raises ValidationError if not found.
+        :param bin_name: The name of the binary executable.
+        :return: None.
+        """
+        if shutil.which(bin_name) is None:
+            raise ValidationError(
+                self.name,
+                f"Can't find {bin_name} executable. Please make sure it's installed.",
+            )
+
+    def _assert_version_in_range(
+        self, app_name: str, version: str, min_version: str, max_version: str
+    ) -> None:
+        """
+        Checks if the given app_name with a string version falls in between specified min and max
+        versions (min_version <= version < max_version). Raises ValidationError.
```

Comment:
Not sure, maybe `max_version` should also be included? Or be renamed?

Reply:
I'll add `_exc`